File: 037/cpp/main.cpp
class Solution {
public:
    vector<tuple<int, int>> location(vector<vector<char>> &board) {
        vector<tuple<int, int>> t;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    t.push_back({i, j});
                }
            }
        }
        return t;
    }

    bitset<9> find(vector<vector<char>> &board, int i, int j) {
        bitset<9> b;
        for (int k = 0; k < 9; k++) {
            char c = board[i][k];
            if (c == '.') {
                continue;
            }
            b.set(c - '0' - 1);
        }
        for (int k = 0; k < 9; k++) {
            char c = board[k][j];
            if (c == '.') {
                continue;
            }
            b.set(c - '0' - 1);
        }
        for (int k = 0; k < 9; k++) {
            char c = board[(i / 3) * 3 + k / 3][(j / 3) * 3 + k % 3];
            if (c == '.') {
                continue;
            }
            b.set(c - '0' - 1);
        }
        b.flip();
        return b;
    }

    bool solve(vector<vector<char>> &board, vector<tuple<int, int>> &t, int k) {
        if (t.size() <= k) {
            return true;
        }
        auto [i, j] = t[k];
        auto b = find(board, i, j);
        for (int n = 0; n < 9; n++) {
            if (!b.test(n)) {
                continue;
            }
            board[i][j] = n + 1 + '0';
            if (solve(board, t, k + 1)) {
                return true;
            } 
        }
        board[i][j] = '.';
        return false;
    }

    void solveSudoku(vector<vector<char>>& board) {
        auto t = location(board);
        solve(board, t, 0);
    }
};

File: 037/rust/main.rs
use std::collections::HashSet;

impl Solution {
    fn location(board: &Vec<Vec<char>>) -> Vec<(usize, usize)> {
        let mut t = Vec::new();
        for i in 0..9 {
            for j in 0..9 {
                if board[i][j] == '.' {
                    t.push((i, j));
                }
            }
        }
        return t;
    }

    fn find(board: &Vec<Vec<char>>, i: usize, j: usize) -> HashSet<char> {
        let mut s1 = HashSet::new();
        for k in 0..9 {
            let c = board[i][k];
            if c == '.' {
                continue;
            }
            s1.insert(c);
        }
        for k in 0..9 {
            let c = board[k][j];
            if c == '.' {
                continue;
            } 
            s1.insert(c);
        }
        for k in 0..9 {
            let c = board[(i / 3) * 3 + k / 3][(j / 3) * 3 + k % 3];
            if c == '.' {
                continue;
            }
            s1.insert(c);
        }

        let mut s2 = HashSet::new();
        for c in vec!['1', '2', '3', '4', '5', '6', '7', '8', '9'] {
            if !s1.contains(&c) {
                s2.insert(c);
            }
        }
        return s2;
    }

    pub fn solve(board: &mut Vec<Vec<char>>, location: &mut Vec<(usize, usize)>, k: usize) -> bool {
        if location.len() <= k {
            return true;
        }
        let (i, j) = location[k];
        let s = Solution::find(board, i, j);
        for c in s {
            board[i][j] = c;
            if Solution::solve(board, location, k + 1) {
                return true;
            }
        }
        board[i][j] = '.';
        return false;
    }

    pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
        let mut location = Solution::location(board);
        Solution::solve(board, &mut location, 0);
    }
}
